### ISSUE TITLE
Unable to pull new images

### ISSUE THREAD

**Issue by cgallimore** [chef-cookbooks/docker#547]
Docker 1.9.0 and Docker cookbook 2.2.3

         Docker::Error::ArgumentError
          ----------------------------
          Must have id, got: {"id"=>nil, :headers=>{"X-Registry-Auth"=>"eyJzZXJ2ZXJhZGRyZXNzIjoicXVheS5pbyIsInVzZXJuYW1lIjoic29uaWFuIiwicGFzc3dvcmQiOiIzeWpwZbmV0In0="}}
          
          Cookbook Trace:
          ---------------
          /var/chef/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.4/lib/docker/base.rb:16:in `initialize'
          /var/chef/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.4/lib/docker/image.rb:117:in `new'
          /var/chef/cache/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.4/lib/docker/image.rb:117:in `create'
          /var/chef/cache/cookbooks/docker/libraries/helpers_image.rb:71:in `block in pull_image'
          /var/chef/cache/cookbooks/docker/libraries/helpers_base.rb:67:in `call'
          /var/chef/cache/cookbooks/docker/libraries/helpers_base.rb:67:in `with_retries'
          /var/chef/cache/cookbooks/docker/libraries/helpers_image.rb:66:in `pull_image'
          /var/chef/cache/cookbooks/docker/libraries/docker_image.rb:61:in `block in <class:DockerImage>'
          /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/provider.rb:23:in `block in action'
          
          Resource Declaration:
          ---------------------
          # In /var/chef/cache/cookbooks/jupiter/recipes/docker.rb
          
           44: docker_image "#{node.docker_registry}/jupiter" do
           45:   tag deploy_tag
           46:   notifies :redeploy, "docker_container[jupiter]", :immediately
           47:   not_if { ::File.exists?(deploy_file) }
           48: end
           49: 
          
          Compiled Resource:
          ------------------
          # Declared in /var/chef/cache/cookbooks/jupiter/recipes/docker.rb:44:in `from_file'
          
          docker_image("repo/jupiter") do
            action [:pull]
            retries 0
            retry_delay 2
            default_guard_interpreter :default
            declared_type :docker_image
            cookbook_name "jupiter"
            recipe_name "docker"
            tag "dev"
            repo "repo/jupiter"
            connection #<Docker::Connection:0x0000000e2deef8 @url="unix:///", @options={:socket=>"/var/run/docker.sock", "read_timeout"=>120}>
            not_if { #code block }
          end

**Comment by cgallimore** [comment]
Looks like this is linked to this Docker-API issue: https://github.com/swipely/docker-api/issues/333

**Comment by someara** [comment]
I've updated kitchen to test against 1.9.0 and was able to converge resources-190-ubuntu-1510 twice without error... reading the linked issue... this is only for re-pulling existing images?

**Comment by cgallimore** [comment]
i've found this to happen even on new instances (in AWS)

**Comment by chasebolt** [comment]
Here is what I think may be happening, though I don't have time to test it.

id is now returning the tag value of an image instead of the image 64 char id. You have an existing image A with a tag of dev, then you are pulling in a new image B with a tag of dev; this process will remove tag dev from image A and now image A has a tag of nil.

We don't have a kitchen test for an image being built, tagged as latest, then the image built again and re-tagged as latest; with an image pull in-between this process. This test should trigger this issue.

**Comment by chasebolt** [comment]
Ok looking deeper into this, I no longer think the above applies to this case (though it is an issue).

@cgallimore Can you verify this isn't the case? https://github.com/swipely/docker-api/issues/315. Check using the cli to verify you can pull the image with the tag. From what it looks like in the trace this is an issue with the image/tag simply not existing.

**Comment by cgallimore** [comment]
@chasebolt i can pull the image using docker pull without problem. tried from a new host and from existing hosts that are seeing this issue.

**Comment by chasebolt** [comment]
are you using a private repo? perhaps a credential issue?

**Comment by cgallimore** [comment]
yes, private repo. credentials haven't changed. the only thing thats changed is updating the base docker cookbook (from a really old pre-1.x version).

**Comment by chasebolt** [comment]
my guess is chef-docker isnt properly authenticated with the repo

**Comment by cgallimore** [comment]
possibly, but the docker_registry resource in my wrapper hasn't changed at all nor has the user/password combo to the repo

**Comment by someara** [comment]
Does this still happen with v2.2.6?

**Comment by someara** [comment]
```ruby
```

**Comment by cgallimore** [comment]
yes. complete repo path. i had truncated it a bit for the original report. 

i've updated to 2.2.7 and now chef will converge all the way, but containers no longer start up.
```
  * docker_container[jupiter] action run[2015-11-12T05:16:42+00:00] INFO: Processing docker_container[jupiter] action run (jupiter::docker line 50)

    - deleting jupiter
    - update jupiter
    -   set force to true (was false)
    - starting jupiter
```
happens every run, the container resource looks like
```
docker_container "jupiter" do
  container_name "jupiter"
  image "repo/jupiter:#{deploy_tag}"
  detach true
  force true
  restart_policy "always"
  action :run
  env "JAVA_OPTS=#{jupiter_memory}g"
  volume ["/etc/jupiter:/etc/jupiter", "/var/log/jupiter:/var/log/jupiter"]
  port ['49152:49152']
end
```

**Comment by someara** [comment]
What happens if you change

```ruby
volume ["/etc/jupiter:/etc/jupiter", "/var/log/jupiter:/var/log/jupiter"]`
```

to

```ruby
binds ["/etc/jupiter:/etc/jupiter", "/var/log/jupiter:/var/log/jupiter"]
```

**Comment by cgallimore** [comment]
Changed the resource to use binds instead. Still get the same behavior, where the container is dropped and restarted every converge.

**Comment by someara** [comment]
@cgallimore can you check if your redeploy problem is fixed in 2.3.5?

**Comment by someara** [comment]
I can't reproduce this... closing due to age. Please re-open if this is still an issue
